std/string.d

```diff
@@ -3080,16 +3137,71 @@ if (isConvertibleToString!Range)
     cast(void) stripRight(ws.byUTF!wchar).array;
 }
 
+/// Ditto
+auto stripRight(Range, Char)(Range str, const(Char)[] chars)
+if (((isBidirectionalRange!Range && isSomeChar!(ElementEncodingType!Range)) ||
+     isConvertibleToString!Range) && isSomeChar!Char)
+{
+    static if (isConvertibleToString!Range)
+        return stripRight!(StringTypeOf!Range)(str, chars);
+    else
+    {
+        for (; !str.empty; str.popBack)
+        {
+            if (chars.indexOf(str.back) == -1)
```

Comment:
Why don't you use `foreach` here?
In this case there's _any_ performance penalty of this abstraction.
https://github.com/dlang/phobos/pull/6092#issuecomment-361397749

Reply:
ditto

Reply:
Here we're iterating from the back, so we'd need `foreach_reverse` or `retro` if you wanted to use an abstraction.

Or we could write this as:
````
while (!str.empty && chars.indexOf(str.back) != -1)
    str.popBack;
````
which to me reads more intuitively as "while the range is not empty and the back is a character to the stripped, pop the back".

Reply:
Shall I change this to use `while` as suggested?

Reply:
Not important, but I like the `for` form. Kurz und gut.

Reply:
> Not important, but I like the for form. Kurz und gut.

Yes, I was referring to `foreach_reverse`:

```d
foreach_reverse (s; str)
    if (chars.indexOf(s) == -1)
        break;
```

FWIW LDC optimizes `foreach_reverse` a lot better:
 
https://godbolt.org/g/k9tr4S

```asm
_D7example3fooFAyaQdZv:
  sub rsp, 72
  mov qword ptr [rsp + 64], rcx
  mov qword ptr [rsp + 56], rdx
  mov qword ptr [rsp + 40], rdi
  mov qword ptr [rsp + 48], rsi
  mov rcx, qword ptr [rsp + 64]
  mov rdx, qword ptr [rsp + 56]
  mov qword ptr [rsp + 24], rdx
  mov qword ptr [rsp + 32], rcx
  mov rcx, qword ptr [rsp + 24]
  mov qword ptr [rsp + 16], rcx
.LBB0_1:
  mov rax, qword ptr [rsp + 16]
  mov rcx, rax
  sub rcx, 1
  mov qword ptr [rsp + 16], rcx
  cmp rax, 0
  je .LBB0_6
  mov eax, 1
  mov rcx, qword ptr [rsp + 16]
  mov rdx, qword ptr [rsp + 32]
  mov sil, byte ptr [rdx + rcx]
  mov byte ptr [rsp + 15], sil
  movzx esi, byte ptr [rsp + 15]
  mov rdx, qword ptr [rsp + 40]
  mov rcx, qword ptr [rsp + 48]
  mov edi, 1
  mov dword ptr [rsp + 8], eax
  call _D3std6string__T7indexOfTAyaZQnFNaNbNiNfQpxwxEQBs8typecons__T4FlagVQBqa13_6361736553656e736974697665ZQBoZl@PLT
  cmp rax, -1
  jne .LBB0_4
  jmp .LBB0_6
.LBB0_4:
  jmp .LBB0_5
.LBB0_5:
  jmp .LBB0_1
.LBB0_6:
  add rsp, 72
  ret
```

I hope you notice that it could inline `back` completely.

For comparison, the `for`:

```asm
_D7example3fooFAyaQdZv:
  push r15
  push r14
  push r12
  push rbx
  sub rsp, 24
  mov r14, rsi
  mov r12, rdi
  mov qword ptr [rsp + 8], rdx
  mov qword ptr [rsp + 16], rcx
  test rdx, rdx
  je .LBB0_9
  mov rdi, rdx
  mov rsi, rcx
  call _D3std5range10primitives__T4backTyaZQjFNaNdNfAyaZw@PLT
  mov edi, 1
  mov esi, eax
  mov rdx, r12
  mov rcx, r14
  call _D3std6string__T7indexOfTAyaZQnFNaNbNiNfQpxwxEQBs8typecons__T4FlagVQBqa13_6361736553656e736974697665ZQBoZl@PLT
  cmp rax, -1
  je .LBB0_9
  lea r15, [rsp + 8]
.LBB0_3:
  mov rbx, qword ptr [rsp + 8]
  test rbx, rbx
  je .LBB0_6
  mov rdi, rbx
  mov rsi, r15
  call _D3std3utf__T10strideBackTAyaZQrFNaNfKQmmZk@PLT
  mov eax, eax
  sub rbx, rax
  cmp rbx, qword ptr [rsp + 8]
  ja .LBB0_5
  mov qword ptr [rsp + 8], rbx
  test rbx, rbx
  je .LBB0_9
  mov rsi, qword ptr [rsp + 16]
  mov rdi, rbx
  call _D3std5range10primitives__T4backTyaZQjFNaNdNfAyaZw@PLT
  mov edi, 1
  mov esi, eax
  mov rdx, r12
  mov rcx, r14
  call _D3std6string__T7indexOfTAyaZQnFNaNbNiNfQpxwxEQBs8typecons__T4FlagVQBqa13_6361736553656e736974697665ZQBoZl@PLT
  cmp rax, -1
  jne .LBB0_3
.LBB0_9:
  add rsp, 24
  pop rbx
  pop r12
  pop r14
  pop r15
  ret
.LBB0_6:
  lea rsi, [rip + .L.str]
  lea rcx, [rip + .L.str.1]
  mov edi, 69
  mov edx, 92
  mov r8d, 2206
  call _d_assert_msg@PLT
.LBB0_5:
  lea rsi, [rip + .L.str.1]
  mov edi, 92
  mov edx, 2207
  call _d_arraybounds@PLT
```

 If we would have perfect abstraction, I would have suggested `str.retro.until!(s => chars.indexOf(s) > 0);`, but unfortunately that's - even for the LLVM optimizer - too much.